### ISSUE TITLE
[Payment history] make refund receipts have correct dates when refunded directly via stripe

### ISSUE THREAD

**Issue by blimpich** [Expensify/App#44127]
# Problem:

Refund date is showing at Jan 1st 1970 on refund receipts.

# Reproduction steps:

1. create new account on Old Dot with an email that you have access to
2. create new annual subscription with a non-zero user count (ex: `5`)
3. add a [valid stripe test card](https://docs.stripe.com/testing#cards) to the user's Account > Payment > Payment Details
4. from the cli, run billing using `script/clitools.sh billingrun`
5. run `php script/notifyall.sh` from inside the VM in Web-Expensify
6. make sure `script/bwm.sh` is running so that the subsequent BillingReceiptUploader job can complete
7. confirm you get a billing receipt
8. comment out [these lines of code from the ProcessStripeWebhook command](https://github.com/Expensify/Auth/blob/main/auth/command/ProcessStripeWebhook.cpp#L52-L55)
9. recompile and restart auth
10. look in the `purchases` table at the most recent purchase (should be the one you just made) and find the `transactionID` value. It'll be a hashed value, not the actual `transactionID`.
11. run the following commands from the VM, careful to replace `YOUR_TRANSACTION_ID_HERE` with your actual transaction id that you got from the `purchases` table
```
nc localhost 4444

ProcessStripeWebhook
eventType: charge.refunded
signature: whsec_8V8HYuNqeMDTgMUdwdGxWygHda7tE2Sx
payload: {"id":"evt_1CA23QKMVuTn5qYVgNZ14TBq","created":1522096592,"data":{"object":{"id":"YOUR_TRANSACTION_ID_HERE","amount":900,"amount_refunded":900,"created":1522096552,"currency":"usd","description":"Refund Test","refunds":{"object":"list","data":[{"id":"re_1CA23QKMVuTn5qYVUgS8I1Mr","object":"refund","amount":900,"balance_transaction":"txn_1CA23QKMVuTn5qYVQXlC0CAx","charge":"ch_1CA22mKMVuTn5qYVQitM5R1L","created":1522096592,"currency":"usd","metadata":[],"reason":"requested_by_customer","receipt_number":null,"status":"succeeded"}],"has_more":false,"total_count":1}}},"type":"charge.refunded"}
countryISO: US
```
12. run `notifyall.sh` again and make sure bwm is still running
13. confirm that you got an email about a refund
14. see that that refund shows Jan 1st 1970 as the refund date

# Solution

Investigate and fix this.

**Comment by dangrous** [comment]
Skimming for issues I can take... Do you still see this happening? Just from the refund that I got for testing that other thing (ha), I see the correct date:

<img width="650" alt="image" src="https://github.com/Expensify/App/assets/5487802/0c744903-d954-4fda-98ce-34d01dc60dc5">

**Comment by blimpich** [comment]
@dangrous huh, not sure! Where did you do the refund from? Was it from supportal or the stripe dashboard? I think the date is only wrong when we refund directly via stripe, which is what the repro steps simulate. But I haven't tried repro-ing since I created this issue so maybe not.

Can you try the repro steps and see if ti reproduces?

**Comment by dangrous** [comment]
oh interesting. Yeah it was via supportal. If I have time today I'll try to go fully through the steps! It was giving me a bit of trouble on Friday but I've rebuilt since then so it's maybe better haha

**Comment by dangrous** [comment]
eyyy finally managed to reproduce. Looking into it now.

**Comment by dangrous** [comment]
Okay so it's this line: https://github.com/Expensify/Auth/blob/ce419dee8d3327b51569f728ce3da131b90f628b/auth/command/ProcessStripeWebhook.cpp#L198

Basically SComposeTime deals with usecs, not regular seconds, so it's not correctly parsing the date response from Stripe (i.e. it's saying 1000x less time has passed).

I just need to figure out the right function to use to parse, I'm sure we have one in there.

**Comment by dangrous** [comment]
Technically can fix by multiplying the purchase date by  `(uint64_t)1000000` but i feel like there's probably a better way

**Comment by trjExpensify** [comment]
Hm, any manual refund requests should get a GH created to process by Conci. Doesn't look like we've had any new ones since this was deployed: https://github.com/search?q=repo%3AExpensify%2FExpensify+refund+request&type=issues

So I think our best bet is to look for refunds in the logs probably, yeah? 

1. Your PR hit prod at ~5.45pm UTC: https://github.com/Expensify/Auth/pull/11590#issuecomment-2229054799
2. When we refund someone, we should trigger `Notification_RefundReceipt` here: https://github.com/Expensify/Web-Expensify/blob/37dbd9df87740a4c4afd91c4edd86862b9090113/lib/UserAPI.php#L3452-L3489
3. I searched the logs for the past two days, and found [this one](https://www.expensify.com/_devportal/tools/logSearch/#query=request_id:(%228a3d186c394b50ad-AKL%22)+AND+timestamp:[2024-07-15T21:12:42.579Z+TO+2024-07-15T23:12:42.579Z]&index=logs_expensify-029399) processed yesterday at 22.12 UTC
4. Supportal'd into the customer account (not incl. credentials here in /app) and can see this receipt in the account: 

<img width="1316" alt="image" src="https://github.com/user-attachments/assets/43cc7171-777d-4d87-8a4c-926a92099349">

5. Not sure why the receipt says refunded on 1st July, 2024 when in fact it was refunded on 15th July, which the billing history in account settings confirms: 

<img width="1077" alt="image" src="https://github.com/user-attachments/assets/63bd7045-5610-4d28-9116-1922d4ef1f6d">

Hope that helps!

**Comment by dangrous** [comment]
Ooh great, that's helpful! But shows that we might be using the wrong date here - the date of the charge, not the date of the refund itself. So, we're close, but no cigar (might not have noticed this testing since the date of the charge and the date of the refund were the same).

@blimpich can you chime in on what we intended here? Asking because we specifically use `    notification["refundDate"] = originalPurchaseDate;` [here](https://github.com/Expensify/Auth/blob/2b65b7e457a801d2033638f077c3707c35d0cbce/auth/lib/Billing.cpp#L1297). Was that intentional? Should the refund date equal the original purchase date? Or is that a separate bug that is now surfaced and needs to be fixed?

**Comment by trjExpensify** [comment]
Yeah, I think it's inaccurate to say "Refunded on X date" and that date not be the date of refund. So we should change that to the date refunded, as it's in the design doc. 👍

**Comment by trjExpensify** [comment]
Worth noting we're adding the card number portion to this as well, so it'll be even weirder to have "Refunded on July 1st 2024 to card ending in 1234"

<img width="783" alt="image" src="https://github.com/user-attachments/assets/1976c1a0-7a2c-4e6f-86aa-cad3ed39e5f8">

**Comment by dangrous** [comment]
Okay cool I'll see if I can spin up a fix for this shortly! I think we can just use `SComposeTime("%Y-%m-%d %H:%M:%S UTC", STimeNow())` - @blimpich would this code ever run at a different time other than immediately after the refund is processed? Trying to figure out if that would be sufficient. I can also see if there's another piece of the data returned from Stripe that I can use.

**Comment by blimpich** [comment]
I think `STimeNow()` is perfectly sufficient 👍

**Comment by dangrous** [comment]
okay that PR is up

**Comment by dangrous** [comment]
Once that's on prod I'll use your test steps @trjExpensify to see if we're good to go!

**Comment by trjExpensify** [comment]
Dope, we're deploying auth now, so shouldn't be too long. 👍

**Comment by blimpich** [comment]
Looks like its deployed 🎉

**Comment by dangrous** [comment]
doesn't look like we've processed a refund yet if I'm calculating UTC correctly. I believe we're shooting for something after July 18 6:40PM UTC and the last one we have is 2024-07-18 16:20:43.

I mean hopefully we never need to refund anyone ever again and we NEVER KNOW

semi-related - I would LOVE a toggle in log search to show times in local time.

**Comment by trjExpensify** [comment]
Move to the UK man.. for half the year you'll be on UTC. 😂

**Comment by dangrous** [comment]
Good/bad news - still no new refunds, so can't confirm nor deny

**Comment by dangrous** [comment]
Okay so I found one - https://www.expensify.com/_devportal/tools/logSearch/#query=request_id:(%22393Ikb%22)+AND+timestamp:[2024-07-22T16:12:42.041Z+TO+2024-07-22T18:12:42.041Z]&index=logs_expensify-029493

Weirdly, the pdf refund is still saying July 1, which is particularly odd because neither the original charge (July 7) nor the refund (July 22) is on that date. I double checked the code though and there's pretty much no way that should have happened, with what I updated. @blimpich do you happen to know if the pdfs are separate code? Like, did I fix it for the email but not the pdf? I didn't realize that if so.

Sorry this is dragging on so much.

**Comment by blimpich** [comment]
The issue is probably that there are three different ways of creating a receipt, and one of them is probably broken. The three ways are:
1. an expensify employee refunds directly via Stripe
2. a expensify employee or support staff (second responders I think) refund via supportal
3. the user themselves refunds via the "request refund" button on the payment page

#2 & #3 use similar logic, but #1 is almost an entirely separate flow. This issue's reproduction steps focus on #1.  I bet that refund was through #2 or #3. We'd need to create a refund locally via supportal or the request refund button to confirm that hypothesis, but that is my guess as to what is going on.

**Comment by dangrous** [comment]
Oh okay I think I've figured it out, mostly - you're right. so for case 1, the notification is sent from from Auth, and we fixed that code. For both 2 ([UserAPI::refundUser](https://github.com/Expensify/Web-Expensify/blob/c3e0e2fb785a2858c8b23506cb7b41797c39c6f1/lib/UserAPI.php#L603)) and 3 ([UserAPI::refundPurchase](https://github.com/Expensify/Web-Expensify/blob/c3e0e2fb785a2858c8b23506cb7b41797c39c6f1/lib/UserAPI.php#L660)), the notification is sent from PHP. Both of those call [UserAPI::createRefundReceipt](https://github.com/Expensify/Web-Expensify/blob/c3e0e2fb785a2858c8b23506cb7b41797c39c6f1/lib/UserAPI.php#L3489) which basically makes the same mistake that we did - they pull the created date of the original purchase, NOT the refund, and use that as the refund date. I think this might be an Auth bug - I think it should be sending back the purchase ID of the refund, not the original purchase.... BUT I'm not sure if there's other code relying on it so to be safe I'll see if we're already sending the refund date somewhere else, and use that rather than changing what Auth sends.

**Comment by dangrous** [comment]
I will have to figure out how to fake those types of refunds on dev though, lemme know if you have tips.

**Comment by dangrous** [comment]
> Weirdly, the pdf refund is still saying July 1, which is particularly odd because neither the original charge (July 7) nor the refund (July 22) is on that date. I double checked the code though and there's pretty much no way that should have happened, with what I updated.

oh and it's always the first of the month when sent from PHP because for some reason we take the date, convert into just a month and year (like `July 2024`) when we send it to the notification, and then change that date BACK to a month day year inside the notification code. But since there's no longer a day in it, it defaults to the first of the month. Le sigh.

**Comment by dangrous** [comment]
Okay I made a draft PR - https://github.com/Expensify/Web-Expensify/pull/42878

Two questions for the room:
* Should I create a new issue? Technically this one is only for "when refunded directly via stripe". We still haven't successfully QAd because we haven't found a matching refund, but the new issue is a separate bug.
* @blimpich I can get the notification code to run successfully (I think) on dev, but it never actually sends the email or creates the refund receipt for view in the app. This happens on `main` as well as my branch, so it's not the new code. Any thoughts on how to test this fully? It's hitting my breakpoints and making the notification, but it never sends, even with BWM` running and after running the notifyall script in the vm.

**Comment by blimpich** [comment]
I think it'd be fine to create a new issue. Probably help keep us all sane. Maybe reword the issue title of this too to make it clear this is just for stripe refunds.

For the second point, hmmmm, that's strange. I'll DM you, maybe we could do some paired programming?

**Comment by dangrous** [comment]
great, made https://github.com/Expensify/Expensify/issues/415121, we'll continue discussion there

As for this, we're still waiting on a Stripe activated refund to QA

**Comment by dangrous** [comment]
Okay so I've found [a Stripe refund](https://www.expensify.com/_devportal/tools/logSearch/#sort=asc&size=10000&query=request_id%3A%228a8506d4ac1276aa-SEA%22%20AND%20timestamp%3A%5B2024-07-18T00%3A00%20TO%202024-07-26T23%3A59%5D), but unfortunately the receipt link in the billing history links to the original receipt, not the refund. Is there another way to look at it?

**Comment by blimpich** [comment]
Did you look at the expenses view? If the refund went through correctly it should have made an expense with the receipt for the refund attached to it.

**Comment by dangrous** [comment]
that one looks like it didn't go through.

BUT (drumroll please) I did find [one](https://www.expensify.com/_devportal/tools/logSearch/#query=request_id:(%228a84249c48c5769a-SEA%22)+AND+timestamp:[2024-07-24T12:09:09.219Z+TO+2024-07-24T14:09:09.219Z]&index=logs_expensify-029521) that DID go through!

<img width="947" alt="Screenshot 2024-07-26 at 13 21 59" src="https://github.com/user-attachments/assets/c2d68875-1be7-40a0-8a28-e0027c74ec46">

Looks like it's working! Closing out.

**Comment by blimpich** [comment]
Oh yay!!!
